### ISSUE TITLE
W1655 incorrectly defined

### ISSUE THREAD

**Issue by wyfo** [PyCQA/pylint#4303]
W1655 states:
> dict.keys referenced when not iterating Used when dict.keys is referenced in a non-iterating context (returns an iterator in Python 3)

However, none of `dict.keys`, `dict.values` and `dict.items` returns an iterator, they return a [dict view](https://docs.python.org/3.8/library/stdtypes.html#dictionary-view-objects).
And `dict.keys` returns a particular view which is "set-like". 

Actually, `dict.keys` is more used (at least in the code I've read) as a set-like object than as a key iterable, because the dict itself is already a key iterable.

See this example:
```python
FORBIDDEN_KEYS: set[str]
def check_data(data: dict[str, Any]):
    if forbidden := (data.keys() & FORBIDDEN_KEYS):  # pylint will raise W1655 here!
        raise ValueError(f"Forbidden keys {forbidden} in data")
```
Pylint raises an error while the use of `dict.keys` is perfectly fine here.

**Comment by hippo91** [comment]
@wyfo thanks for your report. It makes sense.

**Comment by mbyrnepr2** [comment]
Is this issue still valid? I notice W1655 is a deleted message now: https://github.com/PyCQA/pylint/blob/main/pylint/message/_deleted_message_ids.py#L89

**Comment by Pierre-Sassoulas** [comment]
Thank you @mbyrnepr2, you're right. The following slightly modified example (FORBIDDEN_KEYS was undefined and import were lacking):

```python
from __future__ import annotations
from typing import Any

FORBIDDEN_KEYS: set[str] = set()
def check_data(data: dict[str, Any]):
    if forbidden := (data.keys() & FORBIDDEN_KEYS):
        raise ValueError(f"Forbidden keys {forbidden} in data")
```
do not raise anything anymore.
